src/com/linkedin/parseq/internal/Continuations.java

```diff
@@ -44,51 +42,54 @@ public void submit(final Runnable action) {
   }
 
   private static final class Continuation {
-    // contains actions scheduled in the current recurrence level
-    private Deque<Runnable> _active;
+    // contains sibling actions in reverse order of submission
+    // sibling actions are actions submitted by the same parent action
+    private final Deque<Runnable> _siblingActions = new ArrayDeque<>();
 
-    // variable to cache empty deque instance
-    private Deque<Runnable> _inactive;
-
-    // contains actions scheduled for execution
-    private Deque<Runnable> _scheduled;
+    /** contains actions for execution in post-order
+     *  actions with larger depth at the top;
+     *  for sibling actions with the same depth, the first submitted is at the top
+     */
+    private final Deque<Runnable> _postOrderExecutionStack = new ArrayDeque<>();
 
     private void submit(final Runnable action) {
-      if (_active == null) {
+      if (_postOrderExecutionStack.size() == 0) {
         // we are at the root level of a call tree
         // this branch contains main loop responsible for
         // executing all actions
-        _active = _inactive;
-        _scheduled.add(action);
+        _postOrderExecutionStack.push(action);
         loop();
       } else {
-        // not a root level, just schedule an action
-        _active.add(action);
+        // another child action added by the current action
+        _siblingActions.push(action);
       }
     }
 
     private void loop() {
       //  Entering state:
-      //  - _active is empty
-      //  - _scheduled has one action
+      //  - _siblingActions is empty
+      //  - _postOrderExecutionStack has one action (root)
       try {
         do {
-          final Runnable next = _scheduled.pollFirst();
-          next.run();
-          while (!_active.isEmpty()) {
-            _scheduled.addFirst(_active.pollLast());
+          Runnable currentAction = _postOrderExecutionStack.peek();
```

Comment:
If we do pop here, actions submitted by the parent action will not go into _siblingActions (see new line 56).